testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java

```diff
@@ -0,0 +1,342 @@
+                new ReflectPermission("suppressAccessChecks"),
+                new FilePermission(System.getProperty("jboss.inst") + "/standalone/tmp/auth/*", "read")
+        ), "permissions.xml");
+        return archive;
+    }
+
+    public void createDatasource() throws Exception {
+        dsName = java.util.UUID.randomUUID().toString();
+        dsAddress = PathAddress.pathAddress(SUBSYSTEM, "datasources").append("data-source", dsName);
+        dsJndiName = "java:/" + dsName;
+
+        final ModelNode createDsOp = new ModelNode();
+        createDsOp.get(OP).set(ADD);
+        createDsOp.get(OP_ADDR).set(dsAddress.toModelNode());
+        createDsOp.get("jndi-name").set(dsJndiName);
+        createDsOp.get("driver-name").set("h2");
+        createDsOp.get("valid-connection-checker-class-name").set(TestValidConnectionChecker.class.getName());
+        createDsOp.get("valid-connection-checker-module").set("org.jboss.test.customModule");
+        createDsOp.get("exception-sorter-class-name").set(TestExceptionSorter.class.getName());
+        createDsOp.get("exception-sorter-module").set("org.jboss.test.customModule");
+        createDsOp.get("stale-connection-checker-class-name").set(TestStaleConnectionChecker.class.getName());
+        createDsOp.get("stale-connection-checker-module").set("org.jboss.test.customModule");
+        createDsOp.get("min-pool-size").set(2);
+        createDsOp.get("max-pool-size").set(5);
+
+        createDsOp.get("connection-url").set("jdbc:h2:mem:test42");
+        executeAndAssertSuccess(createDsOp);
+
+        dataSource = getDataSourceInstanceFromJndi();
+    }
+
+    public void cleanup() throws Exception {
+        if (connectionList != null) {
+            connectionList.forEach(this::closeIfNecessary);
+            connectionList = null;
+        }
+
+        final ModelNode removeOp = new ModelNode();
+        removeOp.get(OP).set(REMOVE_OPERATION);
+        removeOp.get(OP_ADDR).set(dsAddress.toModelNode());
+        removeOp.get(OPERATION_HEADERS, ModelDescriptionConstants.ALLOW_RESOURCE_SERVICE_RESTART).set(true);
+        executeAndAssertSuccess(removeOp);
+    }
+
+    public void executeAndAssertSuccess(ModelNode op) throws IOException {
+        final ModelNode result = managementClient.getControllerClient().execute(op);
+        if (!result.get(OUTCOME).asString().equals(SUCCESS)) {
+            Assert.fail("Operation failed: " + result.toJSONString(true));
+        }
+    }
+
+    public DataSource getDataSourceInstanceFromJndi() throws NamingException {
+        return (DataSource) new InitialContext().lookup(dsJndiName);
+    }
+
+    public void closeIfNecessary(Connection connection) {
+        try {
+            if (!connection.isClosed()) {
+                connection.close();
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static class Setup implements ServerSetupTask {
```

Comment:
Please use org.jboss.as.test.module.util.TestModule for all this

Reply:
Correcting this is a nice to have; can come later.

Reply:
fixed